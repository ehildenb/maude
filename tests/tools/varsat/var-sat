#!/bin/sh

MAUDE_LIB=$srcdir/../../../src/Main
export MAUDE_LIB

../../../src/Main/maude \
  < var-sat.maude -no-banner -no-advise \
  > var-sat.out 2>&1

diff var-sat.expected var-sat.out > /dev/null 2>&1
