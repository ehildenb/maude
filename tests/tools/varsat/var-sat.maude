

load ../../../contrib/tools/fvp/numbers.maude
load ../../../contrib/tools/varsat/var-sat.maude

fmod TEST-NAT is
  pr VAR-SAT .
  pr EQFORM-OPERATIONS .
  pr NAT       . op nat  : -> Module . vars U V W X Y Z : Nat .
  var F : Form .
  eq nat = upModule('FVP-NAT-PRED, true) .
endfm

red nf-ctor-variants(nat, 'true.Bool != '_<_['X:Nat,'Y:Nat]) .
red toDUPs(getTerms(ctor-variants(hl-func(nat),toHL(nat,FormToTL('true.Bool != '_<_['X:Nat,'Y:Nat]))))) .
red getTerms(ctor-variants(hl-func(nat),toHL(nat,FormToTL('true.Bool != '_<_['X:Nat,'Y:Nat])))) .
red ctor-variants(hl-func(nat),toHL(nat,FormToTL('true.Bool != '_<_['X:Nat,'Y:Nat]))) .
red toHL(nat,FormToTL('true.Bool != '_<_['X:Nat,'Y:Nat])) .

--- Satisfiability
red var-sat(nat, upTerm(X <  Y) ?= 'true.Bool)    == true  .
red var-sat(nat, upTerm(X <  Y) != 'true.Bool)    == true  .
red var-sat(nat, upTerm(X <  X) ?= 'true.Bool)    == false .
red var-sat(nat, upTerm(X <= Y) ?= 'true.Bool)    == true  .
red var-sat(nat, upTerm(X <= X) ?= 'true.Bool)    == true  .
red var-sat(nat, upTerm(X <  Y) ?= 'true.Bool
              /\ upTerm(Y <= Z) ?= 'true.Bool)    == true  .
red var-sat(nat, upTerm(X <  Y) ?= 'true.Bool
              \/ upTerm(Y <= Z) ?= 'true.Bool)    == true  .
red var-sat(nat, upTerm(X <  Y) ?= 'true.Bool
              /\ upTerm(Y <  X) ?= 'true.Bool)    == false .
red var-sat(nat, upTerm(X <  Y) ?= 'true.Bool
              \/ upTerm(Y <  X) ?= 'true.Bool)    == true  .
red var-sat(nat, upTerm(X <  Y) ?= 'true.Bool
              /\ upTerm(Y <  Z) ?= 'true.Bool
              /\ upTerm(Z <  X) ?= 'true.Bool)    == false .

red var-sat(nat, (upTerm(X <  Y       ) ?= 'true.Bool)
              => ('_<_['_+_['X:Nat,'1.NzNat],'Y:Nat] ?= 'true.Bool)) == true .

red var-sat(nat,
       'Z:Bool ?= '_<_['_+_[ 'N:Nat , 'N:Nat ]
                          ,'_+_[ 'M:Nat , 'M:Nat ]]
    /\ 'Z:Bool != 'true.Bool
    /\ 'N:Nat != 'M:Nat)
    == true .

red var-sat(nat,
       'B1:Bool ?= '_<_[ 'N:Nat, 'M:Nat ]
    /\ 'N:Nat != 'M:Nat)
    == true .

red var-valid(nat, (upTerm(X <  Y      ) ?= 'true.Bool)
                => ('_<_['X:Nat,'_+_['Y:Nat,'1.NzNat]] ?= 'true.Bool)) == true .
red var-valid(nat, (upTerm(X <  Y      ) ?= 'true.Bool)
                => ('_<_['_+_['X:Nat,'1.NzNat],'Y:Nat] ?= 'true.Bool)) == false .
red var-valid(nat, (upTerm(X <  Y) ?= 'true.Bool)
                => (upTerm(Y <  X) ?= 'true.Bool)) == false .

reduce var-sat(nat,
        'B2:Bool ?= 'true.Bool
     /\ 'B2:Bool ?= '_<_['N2:Nat,'N1:Nat])
     == true .

reduce var-sat(nat,
            'B2:Bool ?= '_<_['N2:Nat, 'N1:Nat]
         /\ 'N2:Nat ?= 'N1:Nat)
    == true .

reduce var-sat(nat,
            'B2:Bool ?= 'true.Bool
         /\ 'B2:Bool ?= '_<_[ 'N2:Nat, 'N1:Nat]
         /\ 'N2:Nat ?= 'N1:Nat)
    == false .

reduce var-sat(nat,
                'B1:Bool ?= '_<_[ 'N1:Nat, 'N2:Nat  ]
             /\ 'N1:Nat != 'N2:Nat )
    == true .

reduce var-sat(nat,
            'B2:Bool ?= 'true.Bool
         /\ 'B2:Bool ?= '_<_['N2:Nat,'N1:Nat]
         /\ 'N2:Nat != 'N1:Nat)
      == true .

reduce var-sat(nat,
        'B1:Bool ?= '_<_['N2:Nat, '_+_['N1:Nat,'N1:Nat]]
     /\ 'B1:Bool != 'true.Bool)
     == true .


--- These checks are to make sure we work with the theories that nelson oppen
--- uses for testing.

fmod MYNATLIST is
  protecting FVP-BOOL-CTOR .
  sort Nat .

  op fake-zero :     -> Nat [ctor] .
  op fake-succ : Nat -> Nat [ctor] .

  sort NeNatList NatList .
  subsort Nat < NeNatList < NatList .
  op __  : NeNatList NeNatList -> NeNatList [ctor assoc] .
  op nil : -> NatList [ctor] .

  op head : NeNatList -> Nat .
  var N : Nat . var L : NeNatList .
  eq head(N)   = N [variant] .
  eq head(N L) = N [variant] .
endfm

fmod TEST-LIST is
  pr VAR-SAT .
  pr EQFORM-OPERATIONS .
  var F : Form .
  op list : -> Module .
  eq list = upModule('MYNATLIST, true) .
endfm

reduce var-sat(list,
                'N1:Nat ?= 'head['L':NeNatList]
             /\ 'N2:Nat ?= 'head['L:NeNatList]
             /\ 'N1:Nat != 'N2:Nat )
    == true .

reduce var-sat(list,
            'B2:Bool ?= 'true.Bool
         /\ 'N2:Nat ?= 'head['L':NeNatList]
         /\ 'N1:Nat ?= 'head['L:NeNatList]
         /\ 'L':NeNatList ?= 'L:NeNatList
         /\ 'N2:Nat != 'N1:Nat
        )
    == false .

reduce var-sat(list,
            'N2:Nat ?= 'head['L':NeNatList]
         /\ 'N1:Nat ?= 'head['L:NeNatList]
         /\ 'B1:Bool != 'true.Bool
         /\ 'N2:Nat != 'N1:Nat)
      == true .

reduce var-sat(list,
        'I1:Nat ?= 'head['L:NeNatList]
     /\ 'I2:Nat ?= 'head['M:NeNatList]
     /\ 'I1:Nat != 'I2:Nat)
     == true .

reduce var-sat(list,
        'I1:Nat ?= 'I2:Nat
     /\ 'I1:Nat ?= 'head['L:NeNatList]
     /\ 'I2:Nat ?= 'head['M:NeNatList])
     == true .

reduce var-sat(list,
        'N2:Nat ?= 'head['L':NeNatList]
     /\ 'N1:Nat ?= 'head['L:NeNatList]
     /\ 'B1:Bool != 'true.Bool)
     == true .

reduce var-sat(list,
        'I1:Nat ?= 'head['L:NeNatList]
     /\ 'I2:Nat ?= 'head['M:NeNatList])
     == true .

reduce var-sat(list,
        'B2:Bool ?= 'true.Bool
     /\ 'N2:Nat ?= 'head['L':NeNatList]
     /\ 'N1:Nat ?= 'head['L:NeNatList]
     /\ 'L':NeNatList ?= 'L:NeNatList)
     == true .

reduce var-sat(list,
        'I1:Nat ?= 'head['L:NeNatList]
     /\ 'X:Nat  != 'head['L:NeNatList]
     /\ 'Y:Nat  != 'head['L:NeNatList])
     == true .

reduce var-sat(list,
        'I1:Nat ?= 'head['L:NeNatList]
     /\ 'I1:Nat != 'X:Nat
     /\ 'X:Nat  != 'head['L:NeNatList]
     /\ 'Y:Nat  != 'head['L:NeNatList])
     == true .

reduce var-sat(list,
        'I1:Nat ?= 'head['L:NeNatList]
     /\ 'I1:Nat != 'Y:Nat
     /\ 'X:Nat != 'head['L:NeNatList]
     /\ 'Y:Nat != 'head['L:NeNatList])
     == true .

reduce var-sat(list,
        'I1:Nat ?= 'head['L:NeNatList]
     /\ 'X:Nat != 'Y:Nat
     /\ 'X:Nat != 'head['L:NeNatList]
     /\ 'Y:Nat != 'head['L:NeNatList])
     == true .

reduce var-sat(list,
        'I1:Nat ?= 'head['L:NeNatList]
     /\ 'I1:Nat != 'X:Nat
     /\ 'I1:Nat != 'Y:Nat
     /\ 'X:Nat != 'Y:Nat
     /\ 'X:Nat != 'head['L:NeNatList]
     /\ 'Y:Nat != 'head['L:NeNatList])
     == true .

reduce var-sat(list,
        'I1:Nat ?= 'X:Nat
     /\ 'I1:Nat ?= 'head['L:NeNatList]
     /\ 'X:Nat != 'head['L:NeNatList]
     /\ 'Y:Nat != 'head['L:NeNatList])
     == false .

reduce var-sat(list,
        'I1:Nat ?= 'Y:Nat
     /\ 'I1:Nat ?= 'head['L:NeNatList]
     /\ 'X:Nat != 'head['L:NeNatList]
     /\ 'Y:Nat != 'head['L:NeNatList])
     == false .

reduce var-sat(list,
        'I1:Nat ?= 'head['L:NeNatList]
     /\ 'X:Nat ?= 'Y:Nat
     /\ 'X:Nat != 'head['L:NeNatList]
     /\ 'Y:Nat != 'head['L:NeNatList])
     == true
     .
