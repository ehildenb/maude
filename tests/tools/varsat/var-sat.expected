==========================================
reduce in TEST-NAT : addDecls(emptyFModule, setRls(nat, none)) == nat .
rewrites: 39
result Bool: false
==========================================
reduce in TEST-NAT : fin-sorts(nat, 'true.Bool != '_<_['X:Nat,'Y:Nat], none,
    none) .
rewrites: 701
result EmptyTypeSet: (none).EmptyTypeSet
==========================================
reduce in TEST-NAT : var-sat-disj(nat, none, 'true.Bool != '_<_['X:Nat,'Y:Nat])
    .
rewrites: 1378
result Bool: true
==========================================
reduce in TEST-NAT : var-sat-conj(nat, none, 'true.Bool != '_<_['X:Nat,'Y:Nat])
    .
rewrites: 177
result Bool: true
==========================================
reduce in TEST-NAT : var-sat-opt(addDecls(emptyFModule, setRls(nat, none)),
    fin-sorts(nat, 'true.Bool != '_<_['X:Nat,'Y:Nat], none, none), 'true.Bool
    != '_<_['X:Nat,'Y:Nat]) .
rewrites: 1477
result Bool: true
==========================================
reduce in TEST-NAT : hl-func(nat) .
rewrites: 3
result FModule: fmod 'FVP-NAT-PRED is
  nil
  sorts '@HeterogeneousList ; 'Bool ; 'Nat ; 'NzNat .
  subsort 'NzNat < 'Nat .
  op '0 : nil -> 'Nat [ctor] .
  op '1 : nil -> 'NzNat [ctor] .
  op '@S : 'Bool -> '@HeterogeneousList [ctor] .
  op '@S : 'Nat -> '@HeterogeneousList [ctor] .
  op '_+_ : 'Nat 'Nat -> 'Nat [assoc comm ctor id('0.Nat)] .
  op '_+_ : 'Nat 'NzNat -> 'NzNat [assoc comm ctor id('0.Nat)] .
  op '_<=_ : 'Nat 'Nat -> 'Bool [none] .
  op '_<_ : 'Nat 'Nat -> 'Bool [none] .
  op '_monus_ : 'Nat 'Nat -> 'Nat [none] .
  op '_|_ : '@HeterogeneousList '@HeterogeneousList -> '@HeterogeneousList [
    assoc ctor] .
  op 'false : nil -> 'Bool [ctor special(
    id-hook('SystemFalse, nil))] .
  op 'max : 'Nat 'Nat -> 'Nat [comm] .
  op 'min : 'Nat 'Nat -> 'Nat [comm] .
  op 's : 'Nat -> 'Nat [none] .
  op 'sd : 'Nat 'Nat -> 'Nat [comm] .
  op 'singleton : '@HeterogeneousList -> '@HeterogeneousList [ctor] .
  op 'true : nil -> 'Bool [ctor special(
    id-hook('SystemTrue, nil))] .
  none
  eq '_<=_['N:Nat,'_+_['N:Nat,'N':Nat]] = 'true.Bool [variant] .
  eq '_<=_['_+_['N:Nat,'NzN:NzNat],'N:Nat] = 'false.Bool [variant] .
  eq '_<_['N:Nat,'_+_['N:Nat,'NzN:NzNat]] = 'true.Bool [variant] .
  eq '_<_['_+_['N:Nat,'N':Nat],'N:Nat] = 'false.Bool [variant] .
  eq '_monus_['N:Nat,'_+_['N:Nat,'M:Nat]] = '0.Nat [variant] .
  eq '_monus_['_+_['N:Nat,'M:Nat],'N:Nat] = 'M:Nat [variant] .
  eq 'max['N:Nat,'_+_['N:Nat,'M:Nat]] = '_+_['N:Nat,'M:Nat] [variant] .
  eq 'min['N:Nat,'_+_['N:Nat,'M:Nat]] = 'N:Nat [variant] .
  eq 's['N:Nat] = '_+_['1.NzNat,'N:Nat] [variant] .
  eq 'sd['N:Nat,'_+_['N:Nat,'M:Nat]] = 'M:Nat [variant] .
endfm
==========================================
reduce in TEST-NAT : toHL(nat, FormToTL('true.Bool != '_<_['X:Nat,'Y:Nat])) .
rewrites: 33
result Term: '_|_['@S['true.Bool],'@S['_<_['X:Nat,'Y:Nat]]]
==========================================
reduce in TEST-NAT : consistent?(nat, 'Bool, 'false.Bool != 'true.Bool |
    'true.Bool != 'true.Bool) .
rewrites: 18
result Bool: true
==========================================
reduce in TEST-NAT : consistent?(nat, 'Bool, tt | 'false.Bool != 'true.Bool) .
rewrites: 19
result Bool: true
==========================================
reduce in TEST-NAT : consistent?(nat, 'Bool, ff | 'false.Bool != 'true.Bool) .
rewrites: 3
result Bool: false
==========================================
reduce in TEST-NAT : consistent?(nat, 'Bool, tt | ff | 'false.Bool !=
    'true.Bool) .
rewrites: 4
result Bool: false
==========================================
reduce in TEST-NAT : nf-ctor-variants(nat, 'true.Bool != '_<_['X:Nat,'Y:Nat]) .
rewrites: 164
result NegEqLitSet: 'false.Bool != 'true.Bool | 'true.Bool != 'true.Bool
==========================================
reduce in TEST-NAT : toDUPs(getTerms(ctor-variants(hl-func(nat), toHL(nat,
    FormToTL('true.Bool != '_<_['X:Nat,'Y:Nat]))))) .
rewrites: 159
result NegEqLitSet: 'false.Bool != 'true.Bool | 'true.Bool != 'true.Bool
==========================================
reduce in TEST-NAT : getTerms(ctor-variants(hl-func(nat), toHL(nat, FormToTL(
    'true.Bool != '_<_['X:Nat,'Y:Nat])))) .
rewrites: 146
result NeTermSet: ('_|_['@S['true.Bool],'@S['false.Bool]]) | ('_|_['@S[
    'true.Bool],'@S['true.Bool]])
==========================================
reduce in TEST-NAT : ctor-variants(hl-func(nat), toHL(nat, FormToTL('true.Bool
    != '_<_['X:Nat,'Y:Nat]))) .
rewrites: 143
result VariantTripleSet: {'_|_['@S['true.Bool],'@S['false.Bool]],
  'X:Nat <- '_+_['%7:Nat,'%8:Nat] ; 
  'Y:Nat <- '%7:Nat,8,0,false} |
{'_|_['@S['true.Bool],'@S['true.Bool]],
  'X:Nat <- '%4:Nat ; 
  'Y:Nat <- '_+_['%4:Nat,'%5:NzNat],5,0,true}
==========================================
reduce in TEST-NAT : toHL(nat, FormToTL('true.Bool != '_<_['X:Nat,'Y:Nat])) .
rewrites: 33
result Term: '_|_['@S['true.Bool],'@S['_<_['X:Nat,'Y:Nat]]]
==========================================
reduce in TEST-NAT : var-sat(nat, 'true.Bool ?= upTerm(X < Y)) == true .
rewrites: 163
result Bool: true
==========================================
reduce in TEST-NAT : var-sat(nat, 'true.Bool != upTerm(X < Y)) == true .
rewrites: 270
result Bool: true
==========================================
reduce in TEST-NAT : var-sat(nat, 'true.Bool ?= upTerm(X < X)) == false .
rewrites: 112
result Bool: true
==========================================
reduce in TEST-NAT : var-sat(nat, 'true.Bool ?= upTerm(X <= Y)) == true .
rewrites: 163
result Bool: true
==========================================
reduce in TEST-NAT : var-sat(nat, 'true.Bool ?= upTerm(X <= X)) == true .
rewrites: 151
result Bool: true
==========================================
reduce in TEST-NAT : var-sat(nat, 'true.Bool ?= upTerm(X < Y) /\ 'true.Bool ?=
    upTerm(Y <= Z)) == true .
rewrites: 228
result Bool: true
==========================================
reduce in TEST-NAT : var-sat(nat, 'true.Bool ?= upTerm(X < Y) \/ 'true.Bool ?=
    upTerm(Y <= Z)) == true .
rewrites: 182
result Bool: true
==========================================
reduce in TEST-NAT : var-sat(nat, 'true.Bool ?= upTerm(X < Y) /\ 'true.Bool ?=
    upTerm(Y < X)) == false .
rewrites: 147
result Bool: true
==========================================
reduce in TEST-NAT : var-sat(nat, 'true.Bool ?= upTerm(X < Y) \/ 'true.Bool ?=
    upTerm(Y < X)) == true .
rewrites: 173
result Bool: true
==========================================
reduce in TEST-NAT : var-sat(nat, 'true.Bool ?= upTerm(X < Y) /\ 'true.Bool ?=
    upTerm(Z < X) /\ 'true.Bool ?= upTerm(Y < Z)) == false .
rewrites: 181
result Bool: true
==========================================
reduce in TEST-NAT : var-sat(nat, ('true.Bool ?= upTerm(X < Y)) => ('true.Bool
    ?= '_<_['_+_['X:Nat,'1.NzNat],'Y:Nat])) == true .
rewrites: 180
result Bool: true
==========================================
reduce in TEST-NAT : var-sat(nat, 'Z:Bool ?= '_<_['_+_['N:Nat,'N:Nat],'_+_[
    'M:Nat,'M:Nat]] /\ 'M:Nat != 'N:Nat /\ 'Z:Bool != 'true.Bool) == true .
rewrites: 1532
result Bool: true
==========================================
reduce in TEST-NAT : var-sat(nat, 'B1:Bool ?= '_<_['N:Nat,'M:Nat] /\ 'M:Nat !=
    'N:Nat) == true .
rewrites: 1000
result Bool: true
==========================================
reduce in TEST-NAT : var-valid(nat, ('true.Bool ?= upTerm(X < Y)) => (
    'true.Bool ?= '_<_['X:Nat,'_+_['Y:Nat,'1.NzNat]])) == true .
rewrites: 637
result Bool: true
==========================================
reduce in TEST-NAT : var-valid(nat, ('true.Bool ?= upTerm(X < Y)) => (
    'true.Bool ?= '_<_['_+_['X:Nat,'1.NzNat],'Y:Nat])) == false .
rewrites: 709
result Bool: true
==========================================
reduce in TEST-NAT : var-valid(nat, ('true.Bool ?= upTerm(X < Y)) => (
    'true.Bool ?= upTerm(Y < X))) == false .
rewrites: 597
result Bool: true
==========================================
reduce in TEST-NAT : var-sat(nat, 'B2:Bool ?= 'true.Bool /\ 'B2:Bool ?= '_<_[
    'N2:Nat,'N1:Nat]) == true .
rewrites: 208
result Bool: true
==========================================
reduce in TEST-NAT : var-sat(nat, 'B2:Bool ?= '_<_['N2:Nat,'N1:Nat] /\ 'N1:Nat
    ?= 'N2:Nat) == true .
rewrites: 288
result Bool: true
==========================================
reduce in TEST-NAT : var-sat(nat, 'B2:Bool ?= 'true.Bool /\ 'B2:Bool ?= '_<_[
    'N2:Nat,'N1:Nat] /\ 'N1:Nat ?= 'N2:Nat) == false .
rewrites: 177
result Bool: true
==========================================
reduce in TEST-NAT : var-sat(nat, 'B1:Bool ?= '_<_['N1:Nat,'N2:Nat] /\ 'N1:Nat
    != 'N2:Nat) == true .
rewrites: 1000
result Bool: true
==========================================
reduce in TEST-NAT : var-sat(nat, 'B2:Bool ?= 'true.Bool /\ 'B2:Bool ?= '_<_[
    'N2:Nat,'N1:Nat] /\ 'N1:Nat != 'N2:Nat) == true .
rewrites: 1420
result Bool: true
==========================================
reduce in TEST-NAT : var-sat(nat, 'B1:Bool ?= '_<_['N2:Nat,'_+_['N1:Nat,
    'N1:Nat]] /\ 'B1:Bool != 'true.Bool) == true .
rewrites: 536
result Bool: true
==========================================
reduce in TEST-NAT : var-sat(nat, 'false.Bool ?= upTerm(X + X < Y + Y)) .
rewrites: 170
result Bool: true
==========================================
reduce in TEST-NAT : var-sat(nat, 'true.Bool != upTerm(X + X < Y + Y)) .
rewrites: 277
result Bool: true
==========================================
reduce in TEST-LIST : var-sat(list, 'N1:Nat ?= 'head['L':NeNatList] /\ 'N2:Nat
    ?= 'head['L:NeNatList] /\ 'N1:Nat != 'N2:Nat) == true .
rewrites: 2970
result Bool: true
==========================================
reduce in TEST-LIST : var-sat(list, 'B2:Bool ?= 'true.Bool /\ 'N2:Nat ?= 'head[
    'L':NeNatList] /\ 'N1:Nat ?= 'head['L:NeNatList] /\ 'L':NeNatList ?=
    'L:NeNatList /\ 'N1:Nat != 'N2:Nat) == false .
rewrites: 1770
result Bool: true
==========================================
reduce in TEST-LIST : var-sat(list, 'N2:Nat ?= 'head['L':NeNatList] /\ 'N1:Nat
    ?= 'head['L:NeNatList] /\ 'B1:Bool != 'true.Bool /\ 'N1:Nat != 'N2:Nat) ==
    true .
rewrites: 2399
result Bool: true
==========================================
reduce in TEST-LIST : var-sat(list, 'I1:Nat ?= 'head['L:NeNatList] /\ 'I2:Nat
    ?= 'head['M:NeNatList] /\ 'I1:Nat != 'I2:Nat) == true .
rewrites: 1572
result Bool: true
==========================================
reduce in TEST-LIST : var-sat(list, 'I1:Nat ?= 'I2:Nat /\ 'I1:Nat ?= 'head[
    'L:NeNatList] /\ 'I2:Nat ?= 'head['M:NeNatList]) == true .
rewrites: 1116
result Bool: true
==========================================
reduce in TEST-LIST : var-sat(list, 'N2:Nat ?= 'head['L':NeNatList] /\ 'N1:Nat
    ?= 'head['L:NeNatList] /\ 'B1:Bool != 'true.Bool) == true .
rewrites: 1465
result Bool: true
==========================================
reduce in TEST-LIST : var-sat(list, 'I1:Nat ?= 'head['L:NeNatList] /\ 'I2:Nat
    ?= 'head['M:NeNatList]) == true .
rewrites: 1378
result Bool: true
==========================================
reduce in TEST-LIST : var-sat(list, 'B2:Bool ?= 'true.Bool /\ 'N2:Nat ?= 'head[
    'L':NeNatList] /\ 'L':NeNatList ?= 'L:NeNatList /\ 'N1:Nat ?= 'head[
    'L:NeNatList]) == true .
rewrites: 755
result Bool: true
==========================================
reduce in TEST-LIST : var-sat(list, 'I1:Nat ?= 'head['L:NeNatList] /\ 'X:Nat !=
    'head['L:NeNatList] /\ 'Y:Nat != 'head['L:NeNatList]) == true .
rewrites: 1351
result Bool: true
==========================================
reduce in TEST-LIST : var-sat(list, 'I1:Nat ?= 'head['L:NeNatList] /\ 'I1:Nat
    != 'X:Nat /\ 'X:Nat != 'head['L:NeNatList] /\ 'Y:Nat != 'head[
    'L:NeNatList]) == true .
rewrites: 1445
result Bool: true
==========================================
reduce in TEST-LIST : var-sat(list, 'I1:Nat ?= 'head['L:NeNatList] /\ 'I1:Nat
    != 'Y:Nat /\ 'X:Nat != 'head['L:NeNatList] /\ 'Y:Nat != 'head[
    'L:NeNatList]) == true .
rewrites: 1445
result Bool: true
==========================================
reduce in TEST-LIST : var-sat(list, 'I1:Nat ?= 'head['L:NeNatList] /\ 'X:Nat !=
    'Y:Nat /\ 'X:Nat != 'head['L:NeNatList] /\ 'Y:Nat != 'head['L:NeNatList])
    == true .
rewrites: 1449
result Bool: true
==========================================
reduce in TEST-LIST : var-sat(list, 'I1:Nat ?= 'head['L:NeNatList] /\ 'I1:Nat
    != 'X:Nat /\ 'I1:Nat != 'Y:Nat /\ 'X:Nat != 'Y:Nat /\ 'X:Nat != 'head[
    'L:NeNatList] /\ 'Y:Nat != 'head['L:NeNatList]) == true .
rewrites: 1646
result Bool: true
==========================================
reduce in TEST-LIST : var-sat(list, 'I1:Nat ?= 'X:Nat /\ 'I1:Nat ?= 'head[
    'L:NeNatList] /\ 'X:Nat != 'head['L:NeNatList] /\ 'Y:Nat != 'head[
    'L:NeNatList]) == false .
rewrites: 2530
result Bool: true
==========================================
reduce in TEST-LIST : var-sat(list, 'I1:Nat ?= 'Y:Nat /\ 'I1:Nat ?= 'head[
    'L:NeNatList] /\ 'X:Nat != 'head['L:NeNatList] /\ 'Y:Nat != 'head[
    'L:NeNatList]) == false .
rewrites: 2530
result Bool: true
==========================================
reduce in TEST-LIST : var-sat(list, 'I1:Nat ?= 'head['L:NeNatList] /\ 'X:Nat ?=
    'Y:Nat /\ 'X:Nat != 'head['L:NeNatList] /\ 'Y:Nat != 'head['L:NeNatList])
    == true .
rewrites: 1673
result Bool: true
Bye.
